SIG Node contributor ladder review guidance (#6725)
kubernetes/community: sig-node/sig-node-contributor-ladder.md

# SIG Node Contributor Ladder

**Note:**
This document is seeded with content based on initial review from existing SIG Node technical and sub-project leads.

## Goal

> This document intends to define a standard for supporting community members to grow responsibility in the SIG.  The SIG has a goal to scale individual participation using transparent criteria.  It is aspirational, and should be re-evaluated to ensure we can meet the needs of the project with the resources available to participate.  It is expected that it provides the SIG a log for future evolution of decision making criteria.

Reviewer: ```suggestion
This document intends to define a standard for supporting community members to grow responsibility in the SIG.  The SIG has a goal to scale individual participation using transparent criteria.  It is aspirational, and should be re-evaluated periodically to ensure we can meet the needs of the project with the resources available to participate.  It is expected that it provides the SIG a log for future evolution of decision making criteria.
```

If you are a community member that wants to suggest changes, please do!

## New Contributors

> SIG Node welcomes new contributors to the community, help is always desired.  Not all contributors are able to provide sustained contribution, but each contribution is always welcome.  This document intends to capture a leadership path for contributors that intend to provide sustain contribution to the SIG and its associated code base by taking on reviewer and approver responsibilities at various levels.

Reviewer 2: minor spelling error change "provide sustain contribution" to "provide a sustained contribution"

> SIG Node welcomes new contributors to the community, help is always desired.  Not all contributors are able to provide sustained contribution, but each contribution is always welcome.  This document intends to capture a leadership path for contributors that intend to provide sustain contribution to the SIG and its associated code base by taking on reviewer and approver responsibilities at various levels.

Author: great suggestion, added a link.

## SIG Node reviewers and approvers

> SIG Node maintains Kubernetes components that are running large amounts of workloads across clouds, and standalone deployments. Changes in these components have a drastic effect on workload availability. This sets a high bar for SIG Node contributions and puts a lot of emphasis on reliability and security of these changes. Practices and policies of earlier stages of Kubernetes need to be adjusted to the current maturity level of Kubernetes.

Reviewer: I'm a little unclear on what the last sentence "Practices and policies..." is saying here in the context of this doc?

This document explains SIG Node guidelines and requirements to become a reviewer or approver. These requirements are written in the spirit of established Kubernetes [membership document](../community-membership.md) and attempt to help contributors, who want to get reviewer and approver status, to formalize the implicit requirements that the SIG aspires to use when making decisions.

Ultimately, membership in the Kubernetes community, as well as reviewer and approver status is a way to establish trust. Starting with permission to run tests, all the way to deciding on the SIG and Kubernetes overall roadmap for key APIs. No amount of documentation can formalize how this trust is established. This document outlines ways to demonstrate expertise of the code base, sound judgement on decision tradeoffs, end user advocacy, care for community, and ability to work as a distributed team.

> As a high level guidance, SIG Node reviewers and approvers must have a bias to say “no” in reviews and design discussions, and an initial bias towards coding over reviews to demonstrate a baseline for konwledge of code base.  This high level guidance is dictated by the maturity stage of Kubernetes. It is better to have inefficiencies or known bugs than break existing workloads. It is also important to keep the codebase healthy and maintainable.

Reviewer 3: spelling nit: "konwledge" => "knowledge"

> We anticipate providing opportunities for emerging contributors to make these contributions where they express their desire for long term maintenance.  This is facilitated during release planning when pairing approvers with contributors for new feature work and/or e2e health with more testing focused specific contributions as we maintain CI.  Helping maintain our CI and test suite is a great way of demonstrating competency!

Reviewer: ```suggestion
We anticipate providing opportunities for emerging contributors to make these contributions where they express their desire for long term maintenance.  This is facilitated during release planning when pairing approvers with contributors for new feature work and/or specific contributions focused on testing and e2e health as we maintain  our CI.  Helping maintain our CI and test suite is a great way of demonstrating competency!
```

### Reviewer

> Reviewer status is an indication that the person is committed to the SIG activities, technically sound, has accumulated enough context, and is overall trustworthy to help inform approvers and aid contributors. The requirements listed in the [membership document](../community-membership.md#reviewer) highlight this as well.

Reviewer 4: Maybe explicitly comment that you don't have to be a reviewer to add a review / comment on a PR.
Under what circumstances are informal reviews helpful?

Author: Added following:

Reviewer status is an indication that the person is committed to the SIG activities, technically sound, has accumulated enough context, and is overall trustworthy to help inform approvers and aid contributors by applying an lgtm. Anyone is welcome to review a PR with a comment or feedback even if they do not have rights to apply an lgtm.

> Reviewer status is an indication that the person is committed to the SIG activities, technically sound, has accumulated enough context, and is overall trustworthy to help inform approvers and aid contributors. The requirements listed in the [membership document](../community-membership.md#reviewer) highlight this as well.

Reviewer: ```suggestion
Reviewer status is an indication that the person is committed to the SIG activities, demonstrates technical depth, has accumulated enough context, and is overall trustworthy to help inform approvers and aid contributors. The requirements listed in the [membership document](../community-membership.md#reviewer) highlight this as well.
```

The following is the breakdown of requirements listed in the membership document broken down to these categories:

- Committed - proof of sustained contributions

> - member for at least 3 months

Reviewer: ```suggestion
- Member for at least 3 months
```
Is this k8s member? or SIG Node member? Would be good to clarify here.

Author: clarified as k8s member.

> - Technically sound

Reviewer: Maybe use the "Demonstrates technical depth" from my suggestion above? A person can't really be technically sound. :)

- Primary reviewer for at least 5 PRs to the codebase
- Reviewed or merged at least 20 substantial PRs to the codebase
- Has enough context
- Knowledgeable about the codebase
- Trustworthy - established trust with the community
- Sponsored by a subproject approver

> - With no objections from other approvers

Reviewer 5: We had one of these briefly in 2020. I'd love us to do the same again, but I'm not sure how effectively we could staff it :(

SIG Node establishes the following clarifications and additional requirements.

**Committed**

The 3 month of activities should be established looking at PRs reviews history.
Active participating in SIG Node or SIG Node CI weekly meetings or other transient meetings that arise when tackling specific problems in future (exceptions are allowed for cases when timezone or other personal limitations are not allowing for the meeting participation)

SIG Node recognizes that 3 months may not be enough to establish a deep understanding of the codebase. Reviewer status for a subfolder is a step towards SIG-wide reviewer. See the areas section of this document below.

> **Technically sound**

Reviewer: ```suggestion
**Demonstrates Technical Depth**
```
feels more descriptive?

Proof of primary reviewership and significant contributions must be provided. Nominees must provide the list of PRs (at least 5 for primary reviewer and 20 substantial PRs authored or reviewed) as suggested in the membership document. Here are additional comments for this list of PRs:

- Reviewed PRs must be merged.
- Since the purpose is to demonstrate the nominee's technical depth, PRs like analyzer warnings fixes, mechanical “find/replace”-type PRs, minor improvements of kubelet logging and insignificant bug fixes are valued, but not counted towards the reviewer status nomination. Lack of reviews of those PRs may be a red flag for nomination approval.

> - Making the most of reviews in a single area is a good indication that a subdirectory reviewer role may be preferred or top-level node reviewer.  Finding the right balance helps balance the nominees future workload.

Reviewer: ```suggestion
- Making the most of reviews in a single area is a good indication that a subdirectory reviewer role may be preferred or top-level node reviewer.  Finding the right balance helps balance the nominee's future workload.
```

- A primary reviewer should drive the review of the PR without significant input / guidance from the approver or other reviewers. 

> **Has enough context**

Reviewer: As a reader, it feels like having context blends in with the technical depth/being "technically sound" above?

Author: Agree, combined the sections.

- Contributions to k8s documentation
- Blog posts - k8s-hosted and external
- Talks at conferences and meetups
- Contributions to other SIGs

**Trustworthy**

Reviewer nominations are accepted by SIG Node approvers. SIG Node approvers take nominations seriously and are invested in building a healthy community.  Nominees should help approvers understand their future goals in the community so we can help continue to build trust and mutual relationships and nurture new opportunities if and when a contributor wants to become an approver!

### Approver

SIG Node approvers have a lot of responsibilities and it is a very demanding role as the project keeps expanding in capability. It is expected that the SIG Node approver keeps the codebase quality high by giving feedback, thoroughly reviewing code, and giving recommendations to SIG members and reviewers.  SIG Node approvers are essentially gatekeepers to keep the code base at high quality. SIG Node maintains a rigidly high bar for becoming a SIG Node approver by developing trust in a community and demonstrating expertise with a bias towards initial code contributions over reviewing PRs.

> SIG Node approver role is not based on a volume of ongoing work, there is no absolute quota. Approver role is based on accumulated trust and knowledge and it is not expected to require 100% full time commitment from any individual member. The expectation is that an active (non emeritus) approver is responsive to a direct ask to review a complex PR or KEP in their area of expertise when solicited.  It is encouraged for a top-level approver to opt out of a review where another approver is more appropriate.

Reviewer 6: Do we want to handle the emeritus approvers handling here as well? At which point do we consider an approver as emeritus?

Author: @saschagrunert i prefer to keep the current PR on how to grow scope.

Reviewer 6: Sounds good!

SIG Node differentiates code approvers and roadmap approvers by keeping the bar of enhancements approver status higher than code approver status.  In turn, achieving approver status for SIG Node is not an all or nothing experience. We require approvers to develop trust by gradually getting this status for subareas or by maintenance in adjacent projects (e.g. container runtimes).

In addition to the formal requirements for the [approver role](../community-membership.md#approver), SIG Node makes these recommendations for nominees for the SIG Node top level approver status on how to demonstrate expertise and develop trust.

A top-level approver MUST have intermediate approver rights in a subfolder OR adjacent vendored project or client-server flow (device plugin, container runtime, etc.).  Ideally approver rights in more than one subfolder before top-level kubelet approver is solicited is desired but not required.  This is a means of earning trust to existing approvers. 

There are different ways to earn trust:

> **Deep expertise across multiplate sub-systems**

Reviewer: ```suggestion
**Deep expertise across multiple sub-systems**
```

- Demonstrated impact across multiple sub-systems.
 - Troubleshooting complex issues that cross-subsystems with deep understanding of tooling and code base.
 - Code contribution informed from low-level analysis or optimization of bottlnecks in kubelet.  Examples would be optimization of pod startup time, materially improving resource allocations, optimizations from pprof analysis, improvements of kubelet to kube-api traffic at scale (lease optimization, etc.).
 - Create and merge major code simplification and/or optimization PRs indicating deep understanding of tradeoffs taken and validation of potential side effects.

**Proficient in features development**

> - Drive a major KEPs at all three stages: 

Reviewer 2: "Drive major KEPs ..." rather than "Drive a major KEPs..."

Reviewer 2: Maybe "Drive a few major KEPs ..."

 - “alpha” - design proposal and discussions
 - “beta” - initial customer feedback collection

>  - “GA/deprecation” - stabilizing feature, following PRRs, or managing deprecation. 

Reviewer 2: Is PRR a thing?  "following PRRs"

>  - Demonstrate ability to stage changes and pass PRRs keeping end users experience and Kubernetes reliability as top priorities.

Reviewer: ```suggestion
 - Demonstrate ability to stage changes and pass PRRs keeping the end user experience and Kubernetes reliability as top priorities.
```

> - Be a reviewer for a few major  KEPs and demonstrate meaningful participation in the review process. 

Reviewer 2: whitespace between "few major__KEPs"

- Drive features in connected projects which have the direct effect on SIG Node (like Runtimes - Containerd or CRI-O, cAdvisor, runc, etc.).
- Give actionable feedback for the KEPs and initial proposals during the SIG node meetings.

**Active community support**

> Be a primary PR reviewer for numerous PRs in multiple areas listed as a requirement for a reviewer.

Reviewer: probably need dashes here to be consistent with other sections.

Actively triage issues and PRs, provide support to contributors to drive their PRs to completion.

**Be present**

> Participate in SIG Node meetings by speaking about KEPs or improvements driven, or find some other way to prove the identity behind GitHub handle.

Reviewer: ```suggestion
Participate in SIG Node meetings by speaking about KEPs or improvements driven, or find some other way to engage SIG member outside of Github activities.
```

Is this closer to what you're looking for? I presume it's about engagement and not verifying identity?

Author: @kikisdeliveryservice it is about knowing there is a real person behind the GitHub handle.  there is some context for this request based on past requests for membership growth.  as a community, we do want to ensure we know the person by some visible engagement in a meeting, meet-up, conference, call, etc.

Reviewer: gotcha, thanks for the clarification!
